test/CMakeLists.txt

```diff
@@ -1,9 +1,17 @@
+include(FindPkgConfig)
+
 llvm_canonicalize_cmake_booleans(SPIRV_SKIP_CLANG_BUILD)
 llvm_canonicalize_cmake_booleans(SPIRV_SKIP_DEBUG_INFO_TESTS)
 
 # required by lit.site.cfg.py.in
 get_target_property(LLVM_SPIRV_DIR llvm-spirv BINARY_DIR)
 set(LLVM_SPIRV_TEST_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+
+# find spirv-val
+pkg_search_module(SPIRV_TOOLS REQUIRED SPIRV-Tools)
```

Comment:
Can we make it optional?

Downstream users might experience problems if SPIRV-Tools is not available in the environment

Reply:
mhh, I think it's fine to require it for tests only. This file is only included if LLVM_INCLUDE_TESTS is set as well.

Reply:
anyway, if we make SPIRV-tools optional we also have to figure out how to make the spirv-val invocations optional inside the tests as well

Reply:
> mhh, I think it's fine to require it for tests only

I guess anyone might launch tests on cloned repo. We probably need to document this dependency in the README

> anyway, if we make SPIRV-tools optional we also have to figure out how to make the spirv-val invocations optional inside the tests as well

We could try to replace `spirv-val` in LIT tests with some substitution which is defined as actual spirv-val binary if it available and as some noop command otherwise

Reply:
> > mhh, I think it's fine to require it for tests only
> 
> I guess anyone might launch tests on cloned repo. We probably need to document this dependency in the README
> 
> > anyway, if we make SPIRV-tools optional we also have to figure out how to make the spirv-val invocations optional inside the tests as well
> 
> We could try to replace `spirv-val` in LIT tests with some substitution which is defined as actual spirv-val binary if it available and as some noop command otherwise

I think that's something we might want to do anyway. Most of the commands executed are the same in a lot of files and only differ in trivial changes (pipe vs temp files).

I think we should rework how we execute tests and have one wrapper for each type of file (spt, ll, cl) and then handle something like that there as well?

Reply:
but then the lit substitution stuff doesn't work anymore. Maybe we should let cmake generate those test headers and then we can make it a little easier to add new tests.

Reply:
@AlexeySachkov how strong is your opinion on making spirv-val optional? Personally I'd say that skipping that if somebody goes through the trouble running those tests then they also should have spirv-val somewhere as it's also not that difficult to setup (especially if we wait until the auto built binaries are available again)

Reply:
I, personally, think that having spirv-val in translator testing is a pretty good idea and this is important. So, I don't want to block this commit

I hope that downstream users will be able to workaround/fulfill this spirv-val requirement - it shouldn't be a very hard to do so

Reply:
I think the most annoying issue will be that a distribution ships an outdated spirv-tools and a deveoper has to grab a newer one and get cmake to use this. I think I will add some documentation on this as most of the changes inside the test/lit.* files I made are due to this reason.

Reply:
> especially if we wait until the auto built binaries are available again

There are binaries available (e.g. https://github.com/KhronosGroup/SPIRV-Tools/releases/download/master-tot/SPIRV-Tools-master-linux-RelWithDebInfo.zip).  It's just that they're not up to date, this one is from July 2018.  I think there is a fair chance that they will actually work for us, even though they are outdated.

Reply:
I am sure they won't work. We use some fairly recent features. I know that I run with my distribution provided spirv-tools into issues where spirv-val wasn't able to parse the spirv due to unknown capabilities and the package version is 2018.4.

The use of SPV_KHR_no_integer_wrap_decoration is one of the reason.

Reply:
not saying it is a good idea, but something like that seems to kind of work: https://github.com/karolherbst/SPIRV-LLVM-Translator/commit/d900d6511c4d29a927644104377f77849d8567e7

I just don't like the idea of adding packages built against a newer version of ubuntu, especially if this includes C++ binaries.

Reply:
And I also don't want to rely on a PPA providing backports...

Reply:
@AlexeySachkov @svenvh maybe we should just download the vulkan sdk and extract the relevant binaries from there? I think this is something we could actually rely on. There are even ubuntu packages.

I will look into this and update the PR if I end up with something reliable and clean.

Reply:
Will resolve this conversation as I think with the vulkan SDK apt repository that's a good enough solution.

Reply:
Sorry for the late feedback. But I have a strong opinion on having dependency on spirv-tools. Actually I think users must have an option to run tests without having to install spirv-tools package. 
Also will Vulkan SDK solution work on Windows OS?

Reply:
Also dependency on spirv-tools will break existing builds in 3rd party build/test systems.

Reply:
> Sorry for the late feedback. But I have a strong opinion on having dependency on spirv-tools. Actually I think users must have an option to run tests without having to install spirv-tools package.
> Also will Vulkan SDK solution work on Windows OS?

there is a vulkan sdk package for windows. But I don't see the point of running tests if you don't run the full tests in the first place, because you don't get the full benefit of running them.

> Also dependency on spirv-tools will break existing builds in 3rd party build/test systems.

yeah.. that's actually a valid concern. Maybe we could make it optional, but print a big warning telling that whatever result the user is getting might not actually be reliable, because there could be potential spirv validation issue, which, personally, I think is kind of the big point here anyway. Although arguably somebody could deal with invalid spirv inside their platform, but they can't expect that to work on others and I think we shouldn't create incentives to be able to just ignore that, which we shouldn't because we don't want patches leading to invalid spirv, right?

So yeah, optional, but printing an easy to catch warning about it?

Reply:
Yes, I'm fine with the warning.
I agree that we don't want patches leading to invalid spirv. Probably we should establish a policy to add spirv-val check for all new tests producing spirv.